optuna/storages/_redis.py

```diff
@@ -365,6 +398,9 @@ def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
             trial.datetime_complete = datetime.now()
             self._redis.set(self._key_trial(trial_id), pickle.dumps(trial))
             self._update_cache(trial_id)
+
```

Comment:
Sorry, after adding LGTM, but it should be better to add:

```
if self.is_heartbeat_supported():
```

Reply:
I think you mean `is_heartbeat_enabled()`, right?

Reply:
Actually, I do not completely agree with your point. Suppose the following scenario:

1. A study is started with heartbeat enabled.
2. A trial is started.
3. The study aborts unexpectedly.
4. The study is reloaded from the storage manually on the console and the running trial is set to FAIL manually.

With the if-statement, the heartbeat would not be deleted, and as such the `_get_stale_trial_ids()` would report the trial, although it is already i state FAIL. I.e. with the if-statement you can go to some inconsistent state.

Reply:
@Turakar Thank you. I understand.
In that case, could you add the following comment?
```
# To ensure that there are no failed trials with heartbeats in the DB under any circumstances
```

After that, I will merge it.

Reply:
Added :slightly_smiling_face: